esmvalcore/preprocessor/_time.py

```diff
@@ -246,6 +296,204 @@ def spans_three_months(time):
+
+    operator: str, optional
+        Select operator to apply.
+        Available operators: 'mean', 'median', 'std_dev', 'variance', 'min',
+        'max'
+
+    Returns
+    -------
+    iris.cube.Cube
+        Annual statistics cube
+    """
+    # TODO: Add weighting in time dimension. See iris issue 3290
+    # https://github.com/SciTools/iris/issues/3290
+
+    operator = get_iris_analysis_operation(operator)
+
+    if not cube.coords('year'):
+        iris.coord_categorisation.add_year(cube, 'time')
+    return cube.aggregated_by('year', operator)
+
+
+def decadal_statistics(cube, operator='mean'):
+    """
+    Compute decadal statistics.
+
+    Note that this function does not weight the decadal mean if
+    uneven time periods are present. Ie, all data inside the decade
+    are treated equally.
+
+    Parameters
+    ----------
+    cube: iris.cube.Cube
+        input cube.
+
+    operator: str, optional
+        Select operator to apply.
+        Available operators: 'mean', 'median', 'std_dev', 'variance', 'min',
+        'max'
+
+    Returns
+    -------
+    iris.cube.Cube
+        Decadal statistics cube
+    """
+    # TODO: Add weighting in time dimension. See iris issue 3290
+    # https://github.com/SciTools/iris/issues/3290
+
+    operator = get_iris_analysis_operation(operator)
+
+    if not cube.coords('decade'):
+        def get_decade(coord, value):
+            """Callback function to get decades from cube."""
+            date = coord.units.num2date(value)
+            return date.year - date.year % 10
+
+        iris.coord_categorisation.add_categorised_coord(
+            cube, 'decade', 'time', get_decade)
+
+    return cube.aggregated_by('decade', operator)
+
+
+def climate_statistics(cube, operator='mean', period='full'):
+    """
+    Compute climate statistics with the specified granularity.
+
+    Computes statistics for the whole dataset. It is possible to get them for
+    the full period or with the data griouped by day, month or season
+
+    Parameters
+    ----------
+    cube: iris.cube.Cube
+        input cube.
+
+    operator: str, optional
+        Select operator to apply.
+        Available operators: 'mean', 'median', 'std_dev', 'variance', 'min',
+        'max'
+
+    period: str, optional
+        Period to compute the statistic over.
+        Available periods: 'full', 'season', 'seasonal', 'monthly', 'month',
+        'mon', 'daily', 'day'
+
+    Returns
+    -------
+    iris.cube.Cube
+        Monthly statistics cube
+    """
+    period = period.lower()
+
+    if period in ('full', ):
+        operator_method = get_iris_analysis_operation(operator)
+        if operator_accept_weights(operator):
+            time_weights = get_time_weights(cube)
+            cube = cube.collapsed(
+                'time', operator_method, weights=time_weights
+            )
+        else:
+            cube = cube.collapsed('time', operator_method)
+        return cube
+
+    if period in ['daily', 'day']:
```

Comment:
It would be better to use the function `_get_period_coord` (defined below) here. Also this fails if the coordinate is already present on the cube.

Reply:
Done!